yarn-project/ethereum/src/l1_tx_utils.ts

```diff
@@ -0,0 +1,389 @@
+          attempts++;
+          const newGasPrice = await this.getGasPrice(
+            gasConfig,
+            attempts,
+            tx.maxFeePerGas && tx.maxPriorityFeePerGas
+              ? { maxFeePerGas: tx.maxFeePerGas, maxPriorityFeePerGas: tx.maxPriorityFeePerGas }
+              : undefined,
+          );
+
+          this.logger?.debug(
+            `L1 Transaction ${currentTxHash} appears stuck. Attempting speed-up ${attempts}/${gasConfig.maxAttempts} ` +
+              `with new priority fee ${formatGwei(newGasPrice.maxPriorityFeePerGas)} gwei`,
+          );
+
+          currentTxHash = await this.walletClient.sendTransaction({
+            ...request,
+            nonce,
+            gas: params.gasLimit,
+            maxFeePerGas: newGasPrice.maxFeePerGas,
+            maxPriorityFeePerGas: newGasPrice.maxPriorityFeePerGas,
+          });
+
+          txHashes.add(currentTxHash);
+          lastAttemptSent = Date.now();
+        }
+        await sleep(gasConfig.checkIntervalMs!);
+      } catch (err: any) {
+        this.logger?.warn(`Error monitoring tx ${currentTxHash}:`, err);
+        if (err.message?.includes('reverted')) {
+          throw err;
+        }
+        await sleep(gasConfig.checkIntervalMs!);
+      }
+      // Check if tx has timed out.
+      if (gasConfig.txTimeoutMs) {
+        txTimedOut = Date.now() - initialTxTime > gasConfig.txTimeoutMs!;
+      }
+    }
+    throw new Error(`L1 Transaction ${currentTxHash} timed out`);
+  }
+
+  /**
+   * Sends a transaction and monitors it until completion
+   * @param request - The transaction request (to, data, value)
+   * @param gasConfig - Optional gas configuration
+   * @returns The receipt of the successful transaction
+   */
+  public async sendAndMonitorTransaction(
+    request: L1TxRequest,
+    gasConfig?: Partial<L1TxUtilsConfig> & { fixedGas?: bigint },
+  ): Promise<TransactionReceipt> {
+    const { txHash, gasLimit } = await this.sendTransaction(request, gasConfig);
+    return this.monitorTransaction(request, txHash, { gasLimit }, gasConfig);
+  }
+
+  /**
+   * Gets the current gas price with bounds checking
+   */
+  private async getGasPrice(
+    _gasConfig?: L1TxUtilsConfig,
+    attempt: number = 0,
+    previousGasPrice?: typeof attempt extends 0 ? never : GasPrice,
+  ): Promise<GasPrice> {
+    const gasConfig = { ...this.config, ..._gasConfig };
+    const block = await this.publicClient.getBlock({ blockTag: 'latest' });
+    const baseFee = block.baseFeePerGas ?? 0n;
+
+    // Get initial priority fee from the network
+    let priorityFee = await this.publicClient.estimateMaxPriorityFeePerGas();
+    let maxFeePerGas = baseFee;
+
+    // Bump base fee so it's valid for next blocks if it stalls
+    const numBlocks = Math.ceil(gasConfig.stallTimeMs! / BLOCK_TIME_MS);
+    for (let i = 0; i < numBlocks; i++) {
+      // each block can go up 12.5% from previous baseFee
+      maxFeePerGas = (maxFeePerGas * (1_000n + 125n)) / 1_000n;
+    }
```

Comment:
Keep in mind that 1559's `maxFeePerGas` is the total fee, which includes both the base and priority fee components. This means that `maxFeePerGas` should be the bumped `baseFee` _plus_ the estimated priority fee. From the 1559 [spec](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1559.md):

```
			# priority fee is capped because the base fee is filled first
			priority_fee_per_gas = min(transaction.max_priority_fee_per_gas, transaction.max_fee_per_gas - block.base_fee_per_gas)
			# signer pays both the priority fee and the base fee
			effective_gas_price = priority_fee_per_gas + block.base_fee_per_gas
			signer.balance -= transaction.gas_limit * effective_gas_price
```

Setting the `maxFeePerGas` to be only the `baseFee` component means that, if the protocol base fee increases consistently and hits this value, then the resulting priority fee will be effectively zero, and no one will pick up the tx.

Reply:
thanks, fixed!